---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%",
  eval = FALSE
)
```

## Shiny server for the metaPR2 database

<!-- badges: start -->
[![DOI](https://zenodo.org/badge/191439796.svg)](https://zenodo.org/badge/latestdoi/191439796)
<!-- badges: end -->


An [interactive database](https://app.metapr2.org/) of eukaryotic metabarcodes compiled from the literature. 

### Presentation

MetaPR2 is a database of published 18S rRNA metabarcodes. This R package launches a shiny application that allows to interact with the database by mapping, searching and downloading the barcodes.

### How run the metaPR2 shiny server

#### Step 1 - Download the R package from GitHub

[https://github.com/vaulot/metapr2-shiny/archive/refs/heads/main.zip](https://github.com/vaulot/metapr2-shiny/archive/refs/heads/main.zip)

#### Step 2 - Install package on your computer

* Unpack Zip file

* Launch shiny_metapr2.Rproj

* (Skip this step) If you are running Windows install Rtools 

  * [https://cran.r-project.org/bin/windows/Rtools/](https://cran.r-project.org/bin/windows/Rtools/)

#### Step 3 - Install the following libraries

```{r}

### Shiny
install.packages("shiny" ) 
install.packages("shinyvalidate" ) # For validation of input (easier to use than shinyFeedback)
install.packages("shinycssloaders" ) # For the spinning wheel
install.packages("shinyWidgets" ) # for Extra widgets - https://dreamrs.github.io/shinyWidgets/index.html
install.packages("markdown" ) # To display text boxes in md
install.packages("DT" ) # For table display

### Tidyverse
install.packages("dplyr" )
install.packages("tidyr" )
install.packages("stringr" )
install.packages("forcats" )

### graphics
install.packages("ggplot2" )
install.packages("viridis" )
install.packages("patchwork" )
install.packages("treemapify" )

### maps
install.packages("leaflet" )
install.packages("leaflet.minicharts" ) # To do nice pie charts

### Misc
install.packages("rio" )


### Bioinformatics

if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("Biostrings")
BiocManager::install("phyloseq")

```


#### Step 4 - Launch the metaPR2 shiny server

Run the following line from R studio


```{r start}

install.packages("devtools")

devtools::load_all(".")

metapr2App()

```

### Panels

* **About**: Basic information on the metapr2 database
* **Datasets**: List of datasets used.
* **Map**: Maps of relative abundance. Taxonomic level can be chosen as well as sample types (RNA/DNA, substrate, depth layer...).
* **Treemap**: Create treemap graphs for samples and taxon selected
* **Query**: Enter a sequence and search for metabarcode that are similar to your sequence using BLAST like search.

### Data processing


* Total read number per sample has been normalized to 100 with 3 decimals so that the value displayed in the different panels correspond to % of total eukaryotic reads.

### Errors

Please report errors in the [Issues page of the metaPR2 primer database](https://xxx).


### Citation

Vaulot, D. et al. (2021). [metaPR2â€¯: An interactive 18S rRNA metabarcode database](). Unpublished

### Resources
* Website: https://app.metapr2.org/
* Docker: https://hub.docker.com/repository/docker/vaulot/metapr2
* Source code: https://github.com/pr2database/metapr2

### Maintainer
* Daniel Vaulot: vaulot@gmail.com

### Contributors

* 

### Versions

1.0.0 - 2021-xx-xx
* Initial version