<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>==================================== • metapr2</title><script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="===================================="><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">metapr2</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-2">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="articles/metapr2.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Background</h6>
    <a class="dropdown-item" href="articles/vignette-data-processing.html">Data processing</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Sidebar</h6>
    <a class="dropdown-item" href="articles/vignette-select.html">Data selection</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Panels</h6>
    <a class="dropdown-item" href="articles/vignette-datasets.html">Datasets</a>
    <a class="dropdown-item" href="articles/vignette-treemaps.html">Treemaps</a>
    <a class="dropdown-item" href="articles/vignette-maps.html">Maps</a>
    <a class="dropdown-item" href="articles/vignette-barplots.html">Barplots</a>
    <a class="dropdown-item" href="articles/vignette-diversity.html">Diversity</a>
    <a class="dropdown-item" href="articles/vignette-query.html">Query</a>
    <a class="dropdown-item" href="articles/vignette-download.html">Download</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-title-body">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>====================================</h1>
      
    </div>


<div id="section" class="section level1">

</div>
<div class="section level1">
<h1 id="to-do---priority">To do - Priority<a class="anchor" aria-label="anchor" href="#to-do---priority"></a></h1>
</div>
<div class="section level1">
<h1 id="section-1">====================================<a class="anchor" aria-label="anchor" href="#section-1"></a></h1>
<div class="section level2">
<h2 id="maps">Maps<a class="anchor" aria-label="anchor" href="#maps"></a></h2>
<ul><li>For ASV map, allow user to select ASV from the left menu and have it as a special color…</li>
<li>Have color ASVs treemap same as Map and barplots treemaps.</li>
<li>Order colors by most abundant to least abundant (only for ASV map)</li>
<li>Color picker - <a href="https://github.com/daattali/colourpicker/" class="external-link uri">https://github.com/daattali/colourpicker/</a>
</li>
<li>Add lines for tropics and poles based on an agreed convention (eg: 33.5N to 33.5S tropical etc). If not, would it be possible to include latitudinal lines across the map?</li>
</ul><div class="section level3">
<h3 id="alpha-and-beta">Alpha and beta<a class="anchor" aria-label="anchor" href="#alpha-and-beta"></a></h3>
</div>
<div class="section level3">
<h3 id="setting-panel">Setting panel<a class="anchor" aria-label="anchor" href="#setting-panel"></a></h3>
<ul><li>Do a new panel with different settings
<ul><li>Allow user to choose palette
<ul><li>Color blind: <a href="https://cran.r-project.org/web/packages/ggokabeito/readme/README.html?s=09" class="external-link uri">https://cran.r-project.org/web/packages/ggokabeito/readme/README.html?s=09</a>
</li>
</ul></li>
<li>Allow use to save selection and reload (format as yaml ?)</li>
</ul></li>
</ul></div>
<div class="section level3">
<h3 id="taxonomy-selection">Taxonomy selection<a class="anchor" aria-label="anchor" href="#taxonomy-selection"></a></h3>
</div>
</div>
<div class="section level2">
<h2 id="dataset-module">Dataset module<a class="anchor" aria-label="anchor" href="#dataset-module"></a></h2>
<div class="section level3">
<h3 id="query">Query<a class="anchor" aria-label="anchor" href="#query"></a></h3>
<ul><li>Allow to select more than one ASV</li>
</ul></div>
<div class="section level3">
<h3 id="other">Other<a class="anchor" aria-label="anchor" href="#other"></a></h3>
<ul><li><p>Write in background taxa selected…</p></li>
<li><p>Order on the legend on the maps the taxa by decreasing abundance in the dataset. Color should follow same order…<br></p></li>
<li><p>Add a treemap of ASVs number.</p></li>
<li><p>Be able to search for a specific species</p></li>
<li><p>Add heatmaps</p></li>
<li><p>Do better feedback: <a href="https://github.com/daattali/shinyalert" class="external-link uri">https://github.com/daattali/shinyalert</a></p></li>
<li><p>Do better disconnect: <a href="https://github.com/daattali/shinydisconnect/" class="external-link uri">https://github.com/daattali/shinydisconnect/</a></p></li>
</ul></div>
</div>
<div class="section level2">
<h2 id="bugs">Bugs<a class="anchor" aria-label="anchor" href="#bugs"></a></h2>
<p>Query. I’m having a strange problem. When I first checked Pterosperma ASV 00244 from the dataset 23 (ASV ID 81431, sequence 03e10e8596dd5f255a475cdf10b01b141a8dc798) in the public database, it found no match. Then you passed me the private link, and I found, of course, within the Amundsen 23, but also in two samples from the OSD and one from Tara. so I went back to the public database to search for it, and now I cannot find matches in any of them, public or private, even in dataset 23 itself.. it is gone. Can you check if you can find it? Sequence: AGCTCCAATAGCGTATATTTAAGTTGTTGCAGTTAAAAAGCTCGTAGTTGGATTTCGGGTGAATGCGACCGGTCCGCCGTATCGGTGTGCACTGGTTGCGTTTATCTTGCTGTCGGGGACGCGCTCCTGGCCTTAACTGGCTGGGACGTGGAGTCGGCGAAGTTACTTTGAAAAAATTAGAGTGTTCAAAGCAGGCCTACGCTCTGAATACATTAGCATGGAATAACGCGATAGGACTCTGGTCTTATTGTGTTGGTCTTCGAGACCGGAGTAATGATTAAGAGGGACAGTTGGGGGCATTCGTATTTCATTGTCAGAGGTGAAATTCTTGGATTTATGAAAGACGAACTTCTGCGAAAGCATTTGCCAAGGATGTTT</p>
</div>
</div>
<div class="section level1">
<h1 id="section-2">====================================<a class="anchor" aria-label="anchor" href="#section-2"></a></h1>
</div>
<div class="section level1">
<h1 id="to-do---latter">To do - Latter<a class="anchor" aria-label="anchor" href="#to-do---latter"></a></h1>
</div>
<div class="section level1">
<h1 id="section-3">====================================<a class="anchor" aria-label="anchor" href="#section-3"></a></h1>
<div class="section level3">
<h3 id="query-1">Query<a class="anchor" aria-label="anchor" href="#query-1"></a></h3>
<ul><li>Add a “Clear” button for Query</li>
</ul></div>
<div class="section level3">
<h3 id="alpha-and-beta-diversity">Alpha and Beta diversity<a class="anchor" aria-label="anchor" href="#alpha-and-beta-diversity"></a></h3>
<ul><li><p>Use vegan</p></li>
<li><p><a href="https://easystats.github.io/parameters/articles/clustering.html" class="external-link uri">https://easystats.github.io/parameters/articles/clustering.html</a></p></li>
<li><p>Use Microbiome package: <a href="https://microbiome.github.io/tutorials/" class="external-link uri">https://microbiome.github.io/tutorials/</a></p></li>
<li><p>Check metabarcoder for nice trees <a href="https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1005404" class="external-link uri">https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1005404</a></p></li>
<li><p>Implement some function of MicrobiomeSeq nnn</p></li>
<li><p>Display output of ordination in box <a href="https://stackoverflow.com/questions/27139348/shiny-printing-console-output-to-a-text-object-without-waiting-for-a-function-t" class="external-link uri">https://stackoverflow.com/questions/27139348/shiny-printing-console-output-to-a-text-object-without-waiting-for-a-function-t</a></p></li>
</ul></div>
<div class="section level3">
<h3 id="time-series">Time series<a class="anchor" aria-label="anchor" href="#time-series"></a></h3>
<ul><li>Use ggphighlight to highligth groups (cf Clarence)</li>
</ul></div>
<div class="section level3">
<h3 id="misc">Misc<a class="anchor" aria-label="anchor" href="#misc"></a></h3>
<ul><li>Allow to add specific private datasets (using same format)</li>
<li>Create splash screen: <a href="https://www.r-bloggers.com/2019/12/shiny-splash-screen-using-modules-and-shinyjs/" class="external-link uri">https://www.r-bloggers.com/2019/12/shiny-splash-screen-using-modules-and-shinyjs/</a>
</li>
<li>Use a dashboard approach: <a href="https://rstudio.github.io/shinydashboard/" class="external-link uri">https://rstudio.github.io/shinydashboard/</a>
</li>
</ul></div>
<div class="section level3">
<h3 id="data-management">Data management<a class="anchor" aria-label="anchor" href="#data-management"></a></h3>
<ul><li>If data become too big switch duckdb (a bit slower but no limitation on size basically…)</li>
<li>It could be read from the Docker file, so do not provide option for Docker file. So create a Docker file version without the private file…</li>
<li>Provide a more detailed tables with more columns for each dataset</li>
<li>Read all the data from the global.R with an error in case of absence of the private dataset. This way data are read only once per process and then not for all the sessions which may cause things to crash if more than 8 users are connecting.</li>
</ul></div>
<div class="section level3">
<h3 id="docker">Docker<a class="anchor" aria-label="anchor" href="#docker"></a></h3>
<ul><li>Use ARGs to have one image without the private dir</li>
<li>Create a compose YAML file</li>
<li>Probably better to have 2 Docker files…. One which will only be used for the Docker repository</li>
</ul></div>
</div>
<div class="section level1">
<h1 id="section-4">==================================================================================================================================<a class="anchor" aria-label="anchor" href="#section-4"></a></h1>
</div>
<div class="section level1">
<h1 id="section-5">==================================================================================================================================<a class="anchor" aria-label="anchor" href="#section-5"></a></h1>
</div>
<div class="section level1">
<h1 id="section-6">====================================<a class="anchor" aria-label="anchor" href="#section-6"></a></h1>
</div>
<div class="section level1">
<h1 id="run-as-package">Run as package<a class="anchor" aria-label="anchor" href="#run-as-package"></a></h1>
</div>
<div class="section level1">
<h1 id="section-7">====================================<a class="anchor" aria-label="anchor" href="#section-7"></a></h1>
<p>devtools::load_all(“.”) # Ctrl + Shift + L to load source all files (/R) and data (/data) metapr2App()</p>
</div>
<div class="section level1">
<h1 id="section-8">====================================<a class="anchor" aria-label="anchor" href="#section-8"></a></h1>
</div>
<div class="section level1">
<h1 id="run-on-google-cloud">Run on Google cloud<a class="anchor" aria-label="anchor" href="#run-on-google-cloud"></a></h1>
</div>
<div class="section level1">
<h1 id="section-9">====================================<a class="anchor" aria-label="anchor" href="#section-9"></a></h1>
<ul><li>Cloud run : <a href="https://metapr2-public-dcx2qwgoua-as.a.run.app/" class="external-link uri">https://metapr2-public-dcx2qwgoua-as.a.run.app/</a>
</li>
<li>Virtual machine: <a href="http://34.126.91.69:3838/metapr2/" class="external-link uri">http://34.126.91.69:3838/metapr2/</a>
</li>
</ul></div>
<div class="section level1">
<h1 id="section-10">====================================<a class="anchor" aria-label="anchor" href="#section-10"></a></h1>
</div>
<div class="section level1">
<h1 id="documentation-with-pkgdown">Documentation with pkgdown<a class="anchor" aria-label="anchor" href="#documentation-with-pkgdown"></a></h1>
</div>
<div class="section level1">
<h1 id="section-11">====================================<a class="anchor" aria-label="anchor" href="#section-11"></a></h1>
<p>usethis::use_pkgdown()</p>
<p>pkgdown::build_site() # Need to restart R</p>
</div>
<div class="section level1">
<h1 id="preview-site">Preview site<a class="anchor" aria-label="anchor" href="#preview-site"></a></h1>
<p>pkgdown::preview_site()</p>
</div>
<div class="section level1">
<h1 id="when-the-readmemd-is-modified-fails-to-work-use-the-build_site-">when the README.md is modified (fails to work, use the build_site )<a class="anchor" aria-label="anchor" href="#when-the-readmemd-is-modified-fails-to-work-use-the-build_site-"></a></h1>
</div>
<div class="section level1">
<h1 id="pkgdownbuild_home">pkgdown::build_home()<a class="anchor" aria-label="anchor" href="#pkgdownbuild_home"></a></h1>
</div>
<div class="section level1">
<h1 id="build-articles-for-the-vignettes-of-the-pr2database-package-web-site">Build articles for the vignettes of the pr2database package web site<a class="anchor" aria-label="anchor" href="#build-articles-for-the-vignettes-of-the-pr2database-package-web-site"></a></h1>
<p>pkgdown::build_articles()</p>
</div>
<div class="section level1">
<h1 id="build-news-when-the-newsmd-is-changed">Build news when the NEWS.md is changed<a class="anchor" aria-label="anchor" href="#build-news-when-the-newsmd-is-changed"></a></h1>
<p>pkgdown::build_news()</p>
</div>
<div class="section level1">
<h1 id="build-reference-when-the-functions-are-modified-need-to-build-documentation-before">Build reference when the functions are modified (need to Build documentation before…)<a class="anchor" aria-label="anchor" href="#build-reference-when-the-functions-are-modified-need-to-build-documentation-before"></a></h1>
<p>pkgdown::build_reference()</p>
</div>
<div class="section level1">
<h1 id="section-12">====================================<a class="anchor" aria-label="anchor" href="#section-12"></a></h1>
</div>
<div class="section level1">
<h1 id="history">History<a class="anchor" aria-label="anchor" href="#history"></a></h1>
</div>
<div class="section level1">
<h1 id="section-13">====================================<a class="anchor" aria-label="anchor" href="#section-13"></a></h1>
<div class="section level2">
<h2 id="week-of-2021-09-01">Week of 2021-09-01<a class="anchor" aria-label="anchor" href="#week-of-2021-09-01"></a></h2>
<ul><li>Adapt Pelago application</li>
<li>Extend to all taxonomic levels</li>
<li>Make functions more general</li>
<li>Go down to asv level</li>
<li>Update size legend</li>
<li>Break server and ui into individual modules (files that are called by source() have to be in a subdirectory so that they are not loaded automatically)</li>
</ul></div>
<div class="section level2">
<h2 id="week-of-2021-09-06">Week of 2021-09-06<a class="anchor" aria-label="anchor" href="#week-of-2021-09-06"></a></h2>
<ul><li>Move application out of golem (too complicated)</li>
<li>Fix bug which was due to fct_taxonomy (need to add "" besides “All” at each level)</li>
<li>Add Panel to select the different data sets</li>
<li>Change favicon</li>
</ul></div>
<div class="section level2">
<h2 id="week-of-2021-09-13">Week of 2021-09-13<a class="anchor" aria-label="anchor" href="#week-of-2021-09-13"></a></h2>
<ul><li>Add Panel “About”</li>
</ul></div>
<div class="section level2">
<h2 id="week-of-2021-09-20">Week of 2021-09-20<a class="anchor" aria-label="anchor" href="#week-of-2021-09-20"></a></h2>
<ul><li>Add download of data in new panel</li>
<li>Transform to a package</li>
<li>BLAST like search</li>
</ul></div>
<div class="section level2">
<h2 id="week-of-2021-09-27">Week of 2021-09-27<a class="anchor" aria-label="anchor" href="#week-of-2021-09-27"></a></h2>
<ul><li>Treemaps</li>
<li>Two types of maps: pie and dominant with a single function</li>
<li>Add phyloseq support (cannot use standard filtering, must use the prune function)</li>
<li>Change way of selecting datasets</li>
<li>Alpha diversity</li>
<li>Beta diversity</li>
<li>Trim low abundance taxa for alpha and beta diversity (use input$)</li>
<li>Move explanations to readme files</li>
<li>NMDS use different distances: Gower, Bray-Curtis, Jensen-Shannon Divergence, Jaccard</li>
</ul></div>
<div class="section level2">
<h2 id="week-of-2021-10-04">Week of 2021-10-04<a class="anchor" aria-label="anchor" href="#week-of-2021-10-04"></a></h2>
<ul><li>Do checks so that if no condition selected no error occurs</li>
<li>Use different X for alpha diversity</li>
<li>Download phyloseq</li>
<li>Move all UI in the main panel to RenderUI on the server side</li>
<li>For the NMDS, use the level below to the selected level for taxonomy plot</li>
<li>Switch from ShinyFeedback to shinyvalidate</li>
<li>Beta diversity - Add other variables for color and shape (temperature, depth etc…)</li>
<li>Barplot using dplyr</li>
<li>Reorder depth_level for barplots</li>
<li>Move all resources needed to /inst/img and inst/readme</li>
<li>Create an app.R in the main directory to load the package and run it.</li>
<li>Creat an “about.md” in the “inst/readme” directory</li>
<li>Histogram as a function of continuous latitude, of temperature, depth… regroup first and then plot</li>
<li>Use tidyevaluation for functions that need character input (barplot)</li>
<li>Do modules (cleaner, better):
<ul><li>Barplot</li>
<li>Phyloseq</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="week-of-2021-10-11">Week of 2021-10-11<a class="anchor" aria-label="anchor" href="#week-of-2021-10-11"></a></h2>
<ul><li>Do modules (cleaner, better):
<ul><li>Maps</li>
<li>Download</li>
<li>Query</li>
<li>Taxonomy (returns taxo())</li>
<li>Datasets</li>
</ul></li>
<li>Change the data handling using the new df_selected.</li>
<li>Handle errors if no data are selected (barplot, maps, phyloseq, treemap)</li>
<li>Update functions with all_of and across without forgetting to use c() or .data[[x]] (to replace !!as.symbol) (see <a href="https://mastering-shiny.org/action-tidy.html" class="external-link uri">https://mastering-shiny.org/action-tidy.html</a>)</li>
<li>Docker
<ul><li>Update dockerignore and Dockerfile</li>
<li>Problem with phyloseq which requires a library to be installed</li>
<li>There was a problem with the fct_sequence file (very mysterious)</li>
<li>Finalize</li>
<li>Push also to Google run</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="week-of-2021-10-18">Week of 2021-10-18<a class="anchor" aria-label="anchor" href="#week-of-2021-10-18"></a></h2>
<ul><li>Change length of asv_code from 8 to 10 (no ambuiguity for the whole database)</li>
<li>Suppress warning messages</li>
<li>Reassign all of metapr2 with PR2 4.14.0 with both DECIPHER and dada2</li>
<li>Create one table for each assignement</li>
<li>Remove the assignement columns from metapr2_asv</li>
<li>Modify dv_utils to take into account this new structure</li>
<li>Update files in Shiny to PR2 4.12</li>
<li>Add back gene region</li>
<li>Remove gene, gene_region, organelle from asv table since this information is already in the dataset table</li>
<li>Change dataset picker</li>
<li>Update GitHub, web site and web app</li>
<li>Remove phyloseq file for the public dataset (phyloseq is about 2 Go once loaded)</li>
<li>Do not compute anymore df_full but build df_selected on the fly to reduce memory use</li>
<li>Cloud run: two version
<ul><li>metapr2 for the class (5 datasets)</li>
<li>metapr2-public with all the datasets from version 1.0</li>
</ul></li>
<li>Build a virtual machine with a Shiny server and R studio (8 Go of Mem and 10 Go disk)
<ul><li>R libraries must be installed as root with sudo</li>
</ul></li>
<li>Use pryr to monitor memory</li>
<li>Remove columns that are not used</li>
<li>Separate download in 2 packets</li>
<li>Style the download buttons</li>
</ul></div>
<div class="section level2">
<h2 id="week-of-2021-10-25">Week of 2021-10-25<a class="anchor" aria-label="anchor" href="#week-of-2021-10-25"></a></h2>
<ul><li>Fix the problem with qs on Docker (cannot read from the library… )</li>
<li>Use trycatch for the 2 cases of data loading</li>
<li>Fix pico_nano in size fraction</li>
<li>Add ecosystem type</li>
<li>Order datasets alphabetically</li>
<li>Fix datasets names</li>
<li>Create 2 docker containers, private and public</li>
<li>Add authentification to restreindre to public datasets: <a href="https://github.com/datastorm-open/shinymanager" class="external-link uri">https://github.com/datastorm-open/shinymanager</a>
</li>
<li>Select data sets based on users</li>
</ul></div>
<div class="section level2">
<h2 id="week-of-2021-11-01">Week of 2021-11-01<a class="anchor" aria-label="anchor" href="#week-of-2021-11-01"></a></h2>
<ul><li>Use blaster package for the query: <a href="https://github.com/manutamminen/blaster" class="external-link uri">https://github.com/manutamminen/blaster</a>
</li>
<li>Move data saving to dvutils</li>
<li>Change user validation</li>
<li>Create map for query
<ul><li>Select row and draw map…</li>
<li><a href="https://rstudio.github.io/DT/shiny.html" class="external-link uri">https://rstudio.github.io/DT/shiny.html</a></li>
<li><a href="https://yihui.shinyapps.io/DT-rows/" class="external-link uri">https://yihui.shinyapps.io/DT-rows/</a></li>
<li>Use the asv_code as row_name</li>
<li>Use selection = ‘single’. Draw map below…</li>
</ul></li>
<li>Finalize query
<ul><li>Clear map when searching</li>
<li>Order by pidf and sum_reads_asv</li>
<li>Format sum_reads_asv with separator space</li>
<li>Remove taxo box on left and asv_min</li>
<li>Do not require fasta_selected use directly asv_set$fasta</li>
<li>Replace asv_all by asv_blast</li>
<li>Say that all asv are searched</li>
<li>Move crosses text just above the map.</li>
<li>Put asv_code in bold</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="week-of-2021-11-08">Week of 2021-11-08<a class="anchor" aria-label="anchor" href="#week-of-2021-11-08"></a></h2>
<ul><li>asv_set is read in module_datasets.R and not in global.R</li>
<li>global is still read into global.R</li>
<li>credentials are saved in a special file credentials.qs</li>
<li>Write readme for barplot (depth limited to 250 m)</li>
<li>barplot: fix error when only one value for a discrete variable</li>
<li>Phyloseq - create function for computing phyloseq on the fly</li>
<li>Remodel alpha and beta diversity for less than 1000 samples
<ul><li>Create one panel with subpanels</li>
<li>Create phyloseq object on the fly</li>
</ul></li>
<li>Do a validation Rmd for metapr2</li>
</ul></div>
<div class="section level2">
<h2 id="week-of-2021-11-15">Week of 2021-11-15<a class="anchor" aria-label="anchor" href="#week-of-2021-11-15"></a></h2>
<ul><li>Major problem fixed in file produced by metapr2_export_asv… Must remove dataset_id from asv_set table because when merging with sample it will remove all the samples for which the dataset_id is higher from the first asv with a given hash value</li>
<li>Remove Fisher distance because it bugs in many cases</li>
<li>Add ecosystem AND salinity for phyloseq functions</li>
<li>Add possibility to change shape for alpha diversity</li>
<li>Phyloseq download (generate PS on the fly if less than 5000 samples)</li>
<li>Create 2 datasets in different directories that are loaded when using different password</li>
<li>asv_set is now a reactive variable. It only needs to be passed to query module so that search is done on full set of ASVs</li>
<li>Now works as a R package.</li>
<li>Create datasets_groups for easy selection
<ul><li>Marine global</li>
<li>Soils</li>
<li>Rivers</li>
<li>Lakes</li>
<li>Polar</li>
</ul></li>
<li>Create a compose.yaml for Docker</li>
<li>Move from metapr2-public to metapr2</li>
</ul><p>=== Version 1.0.0 ===</p>
<ul><li>Create Slack channel</li>
<li>Add ecological function to fasta file</li>
<li>Prevent side panel from scrolling and incorporate logo into side panel</li>
<li>Barplot improvements
<ul><li>Add ecological_function to barplot</li>
<li>Use plotly in</li>
<li>Time series (Year, Month, Date) in the bar plot</li>
</ul></li>
<li>Move github to metapr2-shiny inside pr2database</li>
<li>Use pkg down for help (problem with trying to reference figures!)</li>
</ul></div>
<div class="section level2">
<h2 id="week-of-2021-11-22">Week of 2021-11-22<a class="anchor" aria-label="anchor" href="#week-of-2021-11-22"></a></h2>
<ul><li>Document all panels</li>
<li>Fix small issue in the dataset table (do not replace name when updating selection tool)</li>
<li>Barplots:
<ul><li>Put year boundaries…</li>
<li>Add salinity</li>
</ul></li>
<li>Sidebar. Create scrolling: <a href="https://stackoverflow.com/questions/66582335/dynamically-sized-sidebarpanel-and-mainpanel-in-shiny" class="external-link uri">https://stackoverflow.com/questions/66582335/dynamically-sized-sidebarpanel-and-mainpanel-in-shiny</a>
</li>
<li>Add dataset_id to map label</li>
<li>Update help for sample selection</li>
</ul><p>=== Version 1.0.1 === * Restrict menu datasets to 10 items * Set up vm machine 2 CPU 8 Go e2-standard - 60 per month * Barplots: add number of samples * Diversity Alpha * Use Violin + Sina plot for discrete variable * Allow discretization of continuous variables (e.g. depth, latitude) * Use nginx on Gcloud VM * create back redirection of address * move pelago and micropolar-protists to VM as well</p>
</div>
<div class="section level2">
<h2 id="week-of-2021-11-29">Week of 2021-11-29<a class="anchor" aria-label="anchor" href="#week-of-2021-11-29"></a></h2>
<ul><li>Update barplots for depth allowing all depths to be plotted.</li>
<li>Use … for barplots and phyloseq</li>
<li>Add information about dominant taxon on map popup</li>
<li>Change default for nreads_min to 100</li>
<li>Update server to run as https
<ul><li><a href="https://www.charlesbordet.com/fr/guide-shiny-aws/#6-securiser-application-https" class="external-link uri">https://www.charlesbordet.com/fr/guide-shiny-aws/#6-securiser-application-https</a></li>
<li><a href="https://certbot.eff.org/instructions?ws=nginx&amp;os=ubuntu-20" class="external-link uri">https://certbot.eff.org/instructions?ws=nginx&amp;os=ubuntu-20</a></li>
</ul></li>
<li>Replace “cat” by “message” so that it is registered in the log error</li>
<li>Save/Load settings put on dataset page</li>
<li>Incorporate error management when loading the yaml file</li>
<li>Now the values of the selection are preserved when choosing different datasets</li>
<li>Maps: %&gt;% addProviderTiles(providers$Esri.WorldStreetMap)</li>
</ul></div>
<div class="section level2">
<h2 id="week-of-2021-12-06">Week of 2021-12-06<a class="anchor" aria-label="anchor" href="#week-of-2021-12-06"></a></h2>
<ul><li>Work on taxonomy picker to:
<ul><li>Select more than 1 level</li>
<li>Save and reload taxo</li>
<li>Reset taxonomy</li>
<li>Issues:
<ul><li>The menus are automatically updated when loading new values so values are not update</li>
<li>Solution: Use a flag that do not reset taxonomy and add a button to reset taxonomy</li>
</ul></li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="week-of-2021-12-13">Week of 2021-12-13<a class="anchor" aria-label="anchor" href="#week-of-2021-12-13"></a></h2>
<ul><li>Implement taxonomy picker in the application</li>
<li>Fix different modules
<ul><li>barplot</li>
<li>map</li>
<li>treemap</li>
</ul></li>
<li>Change selection logic in module dataset so that the taxonomy is only that present in the samples considered</li>
<li>Problème for replot buttons</li>
<li>Use instead a validate taxonomy button but with ignoreNull = FALSE so that the plots are created at t=0</li>
</ul></div>
</div>
<div class="section level1">
<h1 id="section-14">====================================<a class="anchor" aria-label="anchor" href="#section-14"></a></h1>
</div>
<div class="section level1">
<h1 id="manage-package">Manage package<a class="anchor" aria-label="anchor" href="#manage-package"></a></h1>
</div>
<div class="section level1">
<h1 id="section-15">====================================<a class="anchor" aria-label="anchor" href="#section-15"></a></h1>
</div>
<div class="section level1">
<h1 id="to-update-namespace-as-well-as-function-rd-files">To update NAMESPACE as well as function Rd files<a class="anchor" aria-label="anchor" href="#to-update-namespace-as-well-as-function-rd-files"></a></h1>
<p>devtools::document()</p>
</div>
<div class="section level1">
<h1 id="to-create-description-only-do-once-">To create DESCRIPTION (only do once !)<a class="anchor" aria-label="anchor" href="#to-create-description-only-do-once-"></a></h1>
<p>usethis::use_description()</p>
</div>
<div class="section level1">
<h1 id="testing-is-only-done-once">Testing is only done once<a class="anchor" aria-label="anchor" href="#testing-is-only-done-once"></a></h1>
<p>usethis::use_testthat() usethis::use_test() # to initialize a basic test file and open it for editing.</p>
</div>
<div class="section level1">
<h1 id="for-licence-done-once">For licence, done once<a class="anchor" aria-label="anchor" href="#for-licence-done-once"></a></h1>
<p>usethis::use_mit_license(copyright_holder = NULL)</p>
<div class="section level2">
<h2 id="update-description-with-package-needed">Update DESCRIPTION with package needed<a class="anchor" aria-label="anchor" href="#update-description-with-package-needed"></a></h2>
<p>This writes the package to the DESCRIPTION file, then the function can be loaded with package::function</p>
<div class="section level3">
<h3 id="shiny">Shiny<a class="anchor" aria-label="anchor" href="#shiny"></a></h3>
<p>usethis::use_package( “pkgload” ) usethis::use_package( “shiny” ) usethis::use_package( “bslib” ) # Bootstrap themes # usethis::use_package( “shinyFeedback” ) # For user feedback usethis::use_package( “shinyvalidate” ) # For validation of input (easier to use than shinyFeedback) usethis::use_package( “shinycssloaders” ) # For the spinning wheel usethis::use_package( “shinyWidgets” ) # for Extra widgets - <a href="https://dreamrs.github.io/shinyWidgets/index.html" class="external-link uri">https://dreamrs.github.io/shinyWidgets/index.html</a> usethis::use_package( “shinymanager” ) # Authentification usethis::use_package( “markdown” ) # To display text boxes in md usethis::use_package( “DT” ) # For table display</p>
</div>
<div class="section level3">
<h3 id="tidyverse">Tidyverse<a class="anchor" aria-label="anchor" href="#tidyverse"></a></h3>
<p>usethis::use_package( “dplyr” ) usethis::use_package( “tidyr” ) usethis::use_package( “stringr” ) usethis::use_package( “forcats” ) usethis::use_package( “lubridate” ) usethis::use_package( “purrr” )</p>
</div>
<div class="section level3">
<h3 id="graphics">graphics<a class="anchor" aria-label="anchor" href="#graphics"></a></h3>
<p>usethis::use_package( “ggplot2” ) usethis::use_package( “viridis” ) usethis::use_package( “patchwork” ) usethis::use_package( “treemapify” ) usethis::use_package( “plotly” ) usethis::use_package( “ggforce” )</p>
</div>
<div class="section level3">
<h3 id="maps-1">maps<a class="anchor" aria-label="anchor" href="#maps-1"></a></h3>
<p>usethis::use_package( “leaflet” ) usethis::use_package( “leaflet.minicharts” ) # To do nice pie charts</p>
</div>
<div class="section level3">
<h3 id="misc-1">Misc<a class="anchor" aria-label="anchor" href="#misc-1"></a></h3>
<p>usethis::use_package( “rio” ) usethis::use_package( “qs” ) # For ultrafast loading usethis::use_package( “pryr” ) # For memory management usethis::use_package( “scrypt” ) # For encrypting</p>
</div>
<div class="section level3">
<h3 id="bioinformatics">Bioinformatics<a class="anchor" aria-label="anchor" href="#bioinformatics"></a></h3>
<p>usethis::use_package( “Biostrings” ) usethis::use_package( “phyloseq” ) usethis::use_package( “blaster” )</p>
</div>
</div>
</div>
<div class="section level1">
<h1 id="for-debugging">For debugging<a class="anchor" aria-label="anchor" href="#for-debugging"></a></h1>
<p>reactlog::reactlog_enable() # CTR-F3 for real time shiny::reactlogShow()</p>
</div>
<div class="section level1">
<h1 id="section-16">====================================<a class="anchor" aria-label="anchor" href="#section-16"></a></h1>
</div>
<div class="section level1">
<h1 id="paper">Paper<a class="anchor" aria-label="anchor" href="#paper"></a></h1>
</div>
<div class="section level1">
<h1 id="section-17">====================================<a class="anchor" aria-label="anchor" href="#section-17"></a></h1>
<ul><li>NAR special web issue: <a href="https://academic.oup.com/nar/pages/submission_webserver" class="external-link uri">https://academic.oup.com/nar/pages/submission_webserver</a> (Dec 20)</li>
</ul></div>
<div class="section level1">
<h1 id="section-18">====================================<a class="anchor" aria-label="anchor" href="#section-18"></a></h1>
</div>
<div class="section level1">
<h1 id="size-of-objects">Size of objects<a class="anchor" aria-label="anchor" href="#size-of-objects"></a></h1>
</div>
<div class="section level1">
<h1 id="section-19">====================================<a class="anchor" aria-label="anchor" href="#section-19"></a></h1>
<p>Full data set….</p>
<p>Object: asv_set - size: 92.86 Mb</p>
<p>Object: asv_set<span class="math inline"><em>d</em><em>f</em> − <em>s</em><em>i</em><em>z</em><em>e</em> : 56.53<em>M</em><em>b</em><em>O</em><em>b</em><em>j</em><em>e</em><em>c</em><em>t</em> : <em>a</em><em>s</em><em>v</em><sub><em>s</em></sub><em>e</em><em>t</em></span>fasta - size: 39.51 Mb Object: global - size: 10.86 Mb</p>
</div>


  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Daniel Vaulot.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.</p>
</div>

    </footer></div>

  

  

  </body></html>

